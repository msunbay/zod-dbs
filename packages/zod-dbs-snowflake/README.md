# zod-dbs-snowflake

Snowflake provider for zod-dbs.

## Notes

- Requires `snowflake-sdk` as a peer dependency.
- You must provide both `database` and `schemaName`.
- Enum types are not yet supported.

## Installation

```bash
npm install zod-dbs-snowflake
```

> Note: This package has a peer dependency on `snowflake-sdk`. If you don't already have it, you might need to install it (depending on your environment, package manager, etc):

```bash
npm install snowflake-sdk
```

## CLI Usage

```bash
npx zod-dbs --provider snowflake \
  --host <account>.snowflakecomputing.com \
  --user <user> --password <password> \
  --account <account> --token <token> \
  --warehouse <warehouse> --role <role> \ # optional
  --database <db> --schema-name <schema>
```

### Provider Specific CLI Options

| Option                   | Description                                                       | Required |
| ------------------------ | ----------------------------------------------------------------- | -------- |
| `--host <host>`          | Snowflake account URL host (e.g., xy12345.snowflakecomputing.com) |          |
| `--account <account>`    | Snowflake account identifier (required)                           | `true`   |
| `--user <user>`          | Username for authentication                                       |          |
| `--password <password>`  | Password for authentication                                       |          |
| `--database <db>`        | Database name to connect to (required)                            | `true`   |
| `--schema-name <schema>` | Schema name to introspect (required)                              | `true`   |
| `--token <token>`        | JWT token for authentication                                      |          |
| `--role <role>`          | Role to assume after connecting                                   |          |
| `--warehouse <name>`     | Virtual warehouse to use                                          |          |

### Configuration File Example

```ts
import { ZodDbsCliConfig } from 'zod-dbs-cli';

// Import needed to load the provider specific configuration types.
import 'zod-dbs-snowflake';

const config: ZodDbsCliConfig = {
  provider: createProvider(),
  account: '<account>',
  token: '<token>',
  warehouse: '<warehouse>', // optional
  role: '<role>', // optional
};

export default config;
```

## Programmatic Usage

```ts
import { generateZodSchemas } from 'zod-dbs';
import { createProvider } from 'zod-dbs-snowflake';

await generateZodSchemas({
  provider: createProvider(),
  config: {
    account: '<account>',
    host: '<account>.snowflakecomputing.com',
    user: '<user>',
    password: '<password>',
    token: '<token>',
    database: '<db>',
    schemaName: '<schema>',
    // optionally: warehouse, role
  },
});
```
