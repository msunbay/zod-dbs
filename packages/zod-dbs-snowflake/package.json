{
  "name": "zod-dbs-snowflake",
  "version": "0.0.7",
  "description": "Snowflake provider for zod-dbs",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test:unit": "vitest --project unit run",
    "test:types": "tsc --noEmit"
  },
  "author": "msunbay",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msunbay/zod-dbs.git",
    "directory": "packages/zod-dbs-snowflake"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "devDependencies": {
    "dotenv": "^17.2.1",
    "snowflake-sdk": "^2.2.0"
  },
  "dependencies": {
    "zod-dbs-core": "workspace:*"
  },
  "peerDependencies": {
    "snowflake-sdk": ">=1.7.0"
  }
}
