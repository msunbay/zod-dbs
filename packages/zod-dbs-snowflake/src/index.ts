import './types.js';

export { createClient } from './client.js';
export { createProvider } from './provider.js';
export { SnowflakeProvider } from './SnowflakeProvider.js';
