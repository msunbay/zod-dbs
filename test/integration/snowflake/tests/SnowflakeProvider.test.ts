import { SnowflakeProvider } from 'zod-dbs-snowflake';

import { getProviderConfig } from '../../utils/context.js';

const IS_CONFIGURED = !!process.env.SNOWFLAKE_HOST;
const describeIntegration = IS_CONFIGURED ? describe : describe.skip;

describeIntegration('SnowflakeProvider', () => {
  it('fetches schema info for provided database/schema', async () => {
    const provider = new SnowflakeProvider();
    const config = getProviderConfig();

    const info = await provider.getSchemaInformation(config);

    expect(info.tables).toBeDefined();
    expect(info).toMatchSnapshot();
  });
});
