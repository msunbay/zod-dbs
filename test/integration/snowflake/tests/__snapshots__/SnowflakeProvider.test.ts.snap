// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SnowflakeProvider (integration) > fetches schema info for provided database/schema 1`] = `
{
  "tables": [
    {
      "columns": [
        {
          "dataType": "NUMBER",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "name": "MENU_ID",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "number",
        },
        {
          "dataType": "NUMBER",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "name": "MENU_TYPE_ID",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "number",
        },
        {
          "dataType": "TEXT",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": 16777216,
          "name": "MENU_TYPE",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "string",
        },
        {
          "dataType": "TEXT",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": 16777216,
          "name": "TRUCK_BRAND_NAME",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "string",
        },
        {
          "dataType": "NUMBER",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "name": "MENU_ITEM_ID",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "number",
        },
        {
          "dataType": "TEXT",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": 16777216,
          "name": "MENU_ITEM_NAME",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "string",
        },
        {
          "dataType": "TEXT",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": 16777216,
          "name": "ITEM_CATEGORY",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "string",
        },
        {
          "dataType": "TEXT",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": 16777216,
          "name": "ITEM_SUBCATEGORY",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "string",
        },
        {
          "dataType": "NUMBER",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "name": "COST_OF_GOODS_USD",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "number",
        },
        {
          "dataType": "NUMBER",
          "defaultValue": undefined,
          "description": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "name": "SALE_PRICE_USD",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "number",
        },
        {
          "dataType": "VARIANT",
          "defaultValue": undefined,
          "description": "Test description",
          "isArray": false,
          "isEnum": false,
          "isNullable": true,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "name": "MENU_ITEM_HEALTH_METRICS_OBJ",
          "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
          "tableName": "MENU",
          "tableType": "table",
          "type": "json",
        },
      ],
      "name": "MENU",
      "schemaName": "MSOLVAAG_LOAD_SAMPLE_DATA_FROM_S3",
      "type": "table",
    },
  ],
}
`;
